File: app/src/main/java/com/osrsd/cache/def/NpcDefinition.java
package com.osrsd.cache.def;

import lombok.Data;

import java.util.HashMap;
import java.util.Map;

@Data
public final class NpcDefinition implements Definition {

    private int id;
    private String name;
    private int size;
    private int[] models;
    private int[] chatheadModels;
    private int standingAnimation;
    private int rotateLeftAnimation;
    private int rotateRightAnimation;
    private int walkingAnimation;
    private int rotate180Animation;
    private int rotate90RightAnimation;
    private int rotate90LeftAnimation;
    private short[] recolorToFind;
    private short[] recolorToReplace;
    private short[] retextureToFind;
    private short[] retextureToReplace;
    private String[] actions;
    private boolean isMinimapVisible;
    private int combatLevel;
    private int widthScale;
    private int heightScale;
    private boolean hasRenderPriority;
    private int ambient;
    private int contrast;
    private int headIcon;
    private int rotationSpeed;
    private int[] configs;
    private int varbitId;
    private int varpIndex;
    private boolean isInteractable;
    private boolean rotationFlag;
    private boolean isPet;
    private Map<Integer, Object> params;

    public NpcDefinition(int id) {
        this.id = id;
        name = "null";
        size = 1;
        standingAnimation = -1;
        rotateLeftAnimation = -1;
        rotateRightAnimation = -1;
        walkingAnimation = -1;
        rotate180Animation = -1;
        rotate90RightAnimation = -1;
        rotate90LeftAnimation = -1;
        actions = new String[5];
        isMinimapVisible = true;
        combatLevel = -1;
        widthScale = 128;
        heightScale = 128;
        headIcon = -1;
        rotationSpeed = 32;
        varbitId = -1;
        varpIndex = -1;
        isInteractable = true;
        rotationFlag = true;
        params = new HashMap<>();
    }

}

File: app/src/main/kotlin/com/osedit/spring/domain/NpcDefinition.kt
package com.osedit.spring.domain

import com.osedit.cache.Config
import com.osedit.cache.Indices
import com.osedit.cache.provider.NpcProvider
import java.nio.ByteBuffer
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.Id

@Entity
data class NpcDefinition(@Id private val id: Int? = 0) : Definition {
    var name: String = "null"
    var size = 1
    var models: IntArray? = null
    var chatheadModels: IntArray? = null
    var standingAnimation = -1
    var rotateLeftAnimation = -1
    var rotateRightAnimation = -1
    var walkingAnimation = -1
    var rotate180Animation = -1
    var rotate90RightAnimation = -1
    var rotate90LeftAnimation = -1
    var recolorToFind: ShortArray? = null
    var recolorToReplace: ShortArray? = null
    var retextureToFind: ShortArray? = null
    var retextureToReplace: ShortArray? = null
    @ElementCollection
    var actions = mutableListOf<String?>(null, null, null, null, null)
    var isMinimapVisible = true
    var combatLevel = -1
    var widthScale = 128
    var heightScale = 128
    var hasRenderPriority = false
    var ambient = 0
    var contrast = 0
    var headIcon = -1
    var rotationSpeed = 32
    @ElementCollection
    var configs: MutableList<Int>? = null
    var varbitId = -1
    var varpIndex = -1
    var isInteractable = true
    var rotationFlag = true
    var isPet = false
    @ElementCollection
    var params = mutableMapOf<Int, String>()

    override fun path(): String {
        return "/npcs"
    }

    override fun config(): Config {
        return Config.NPCS
    }

    override fun indices(): Indices {
        return Indices.CONFIG
    }

    override fun decode(byteBuffer: ByteBuffer): Definition {
        return NpcProvider.decode(byteBuffer, this)
    }
}